### ISSUE TITLE
Usage with MongoDB

### ISSUE THREAD

**Issue by Vitallyz** [awinogrodzki/next-firebase-auth-edge#97]
Sorry, not an issue with the next-firebase-auth-edge but rather a question of how to use it properly in my case. I am new to the backend and especially to authentication flows. 

I am moving from NextAuth to this library since NextAuth has no support for mobile app login. My next step once the web app is done is to get Android and iOS apps to use the same Next.js backend and the same rest API.

I am using MongoDB and with NextAuth I could check the NextAuth's session on the API route and get the userEmail with which I could access all user's data in the database.  Even better I could use midleware.ts to check if the session is valid and forward userEmail to the API routes. API routes will connect to MongoDB and get all the data back to the frontend web app in response. 

I got errors when tried to add Firebase Auth check in middlware.ts and that is how I found this library. 

How can I make this library work in a similar way I used it with NextAuth? 

- Check authentication in the middleware and forward user info (email or uid for example) in the request headers to the API route that the request was made to.

Would it be a good approach or should I use it as in the example, and get user data from the getTokensFromObject in each API route?    

Thank you for your help. 

Sorry, I could not find a better place to post this question.

**Comment by awinogrodzki** [comment]
Hey @Vitallyz! Thanks for the question.

You definitely could do something like this:
```tsx
export async function middleware(request: NextRequest) {
  return authentication(request, {
    loginPath: "/api/login",
    logoutPath: "/api/logout",
    apiKey: authConfig.apiKey,
    cookieName: authConfig.cookieName,
    cookieSerializeOptions: authConfig.cookieSerializeOptions,
    cookieSignatureKeys: authConfig.cookieSignatureKeys,
    serviceAccount: authConfig.serviceAccount,
    handleValidToken: async ({ token, decodedToken }) => {
      // Set custom request headers based on decodedToken <------- THIS SECTION
      
      if (decodedToken?.email) {
        request.headers.set("X-User-Email", decodedToken.email);
      }

      request.headers.set("X-User-Uid", decodedToken.uid);

      // Authenticated user should not be able to access /login, /register and /reset-password routes
      if (PUBLIC_PATHS.includes(request.nextUrl.pathname)) {
        return redirectToHome(request);
      }

      return NextResponse.next();
    },
    handleInvalidToken: async () => {
      return redirectToLogin(request);
    },
    handleError: async (error) => {
      console.error("Unhandled authentication error", { error });
      return redirectToLogin(request);
    },
  });
}
```

And later in API Route:

```tsx
export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  
  const uid = req.headers["x-user-uid"];
  const email = req.headers["x-user-email"];
}
```

**Please note** that in example middleware.ts API routes are not handled by middleware due to this config:
```tsx
// middleware.ts
export const config = {
  matcher: [
    "/",
    "/((?!_next|favicon.ico|api|.*\\.).*)",
    "/api/login",
    "/api/logout",
  ],
};
```

It instructs Next.js to ignore all `/api/*` routes other than `/api/login` and `/api/logout`. That is why all route examples use `getTokensFromObject`. You would have to modify this config to include your api routes.

Now, to answer the question if would it be a good approach:

At the moment API route is called, you already have `uid` or `email` in headers. It's encoded in cookie headers. Let's take this example scenario to show difference between two approaches:

NextAuth way:
1. Browser calls `/api/example-endpoint` route after user was logged in
2. Middleware handles `/api/example-endpoint` route, verifies token and sets additional request headers
3. API route assumes request is authenticated and extracts those additional request headers to identify a user in MongoDB

next-firebase-auth-edge way:
1. Browser calls `/api/example-endpoint` route after user was logged in
2. Middleware **ignores** this route and proceeds request to API route
3. API route calls `getTokensFromObject` which verifies the token and extracts user data from the token


There is no good or bad in both approaches. The difference is where the token get's verified. In NextAuth, it's always verified by middleware. In next-firebase-auth-edge, API routes validate token themselves just before extracting user data.

My motivation for the design was:
1. Not all API routes require authentication, so it's up to API route to decide if it should be authorized
2. I see no point in redirecting user to login via API route, that's why authentication middleware ignores all API routes by default. 

The drawback of 2. is that if there are no user credentials in API route that requires `uid` or `email` we need to handle it explicitly in API route like this:

```tsx
export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  const uid = req.headers["x-user-uid"];
  const email = req.headers["x-user-email"];

  if (!uid) {
    return NextResponse.json({ message: "Not authorized" }, { status: 401 });
  }
  
  // ...success flow
}
```

And now it's up to `/api/example-endpoint` caller to decide if user should be redirected or not.


TLDR; it's perfectly fine to use `getTokenFromObject` in API Route to extract user data before contacting MongoDB, but it's also possible to implement approach similar to `NextAuth` by slightly updating middleware `config` object and `handleValidToken` implementation if needed

**Comment by Vitallyz** [comment]
@awinogrodzki Thank you so much for such a descriptive answer and thanks for the heads up about the matcher! 

Your explanation makes sense thanks for that! 

But it raises another question I had for that library regards user redirect:

In the solved issues I've seen someone complaining about an issue where the backend does not redirect users from register/login/reset-password pages if the user is logged in. Your response was adding the PUBLIC_PATHS array and some logic to redirect users from those pages to the homepage. 

My question is, would it be better to do the redirect thing on the front-end side (and render a fallback view while Firebase is checking auth), since if we go directly to the /login page, it will take some time for the API call to happen, and the login page will be briefly rendered before the redirect to the homepage happens? Or do I miss something in this mechanism?

https://github.com/awinogrodzki/next-firebase-auth-edge/issues/67

Thanks a lot!

**Comment by awinogrodzki** [comment]
@Vitallyz you could implement redirect on front-end side. That way, user with valid credentials would experience something like this:

1. Logged in user opens `/login` page
2. They see the `/login` page for a brief time
3. They get redirected to `/`page

This means at least 3 subsequent requests before authenticated user lands on our target, which is home page.


In middleware-first approach, the one with `PUBLIC_PATHS`, this journey is only 2 request trip:

1. User requests `/login` with authenticated user, middleware intercepts this request, sees that user is authenticated (therefore not allowed to see /login page) and returns response with `307 Temporary Redirect` status and `Location: /` header
2. Browser sees the redirect and immediately redirects user to home page

You can see that with the second approach, we are saving **at least one** network roundtrip. Depending on how many requests are being made in `/login` page.

As I said previously, Next.js is automatically pre-fetching all links, even in `/login` page, so with the second approach we save a considerable amount of network roundtrips. Ofcourse, this would not matter in 90% of the cases, but with scale, this setup can save us a lot of $$$ on AWS bills :-)

**Comment by Vitallyz** [comment]
Thanks, @awinogrodzki I think what I was missing is the fact that it is a server-side rendering I will have to play with this once the auth is working properly in my app.

**Comment by awinogrodzki** [comment]
I will close the issue now. If you have more questions, feel free to create new issue
